### PR TITLE
Switched domain api to elasticsearch

### PR DESCRIPTION
**Pull request by kkrampa**
@dannyroberts

### CODE REVIEW

File: corehq/apps/api/domain_metadata.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
 from datetime import datetime
 import logging
+from tastypie.paginator import Paginator
```
Comment (dimagimon) [inline corehq/apps/api/domain_metadata.py]: F401 'Paginator' imported but unused

**Comment by czue** [comment]
this looks good to me. do you think we need to do any testing? it is just used internally, right?

@esoergel may also want to double check the ES code

File: corehq/apps/api/domain_metadata.py [hunk 2]
```diff
@@ -30,6 +32,21 @@ def get_truth(inp, relate, cut):
...
+    def __init__(self, es_query):
+        self.es_query = es_query
+
+    def count(self):
+        return self.es_query.size(0).run().total
+
+    def __getitem__(self, item):
+        if isinstance(item, slice):
+            doc_ids = self.es_query.start(item.start).size(item.stop - item.start).run().doc_ids
+            return map(Domain.wrap, iter_docs(Domain.get_db(), doc_ids))
```
Comment (snopoke) [inline corehq/apps/api/domain_metadata.py]: Is there a reason we don't just pull the docs directly out of ES?

Comment (kkrampa) [inline corehq/apps/api/domain_metadata.py]: @snopoke Rest of logic in api uses domain objects e.g. https://github.com/dimagi/commcare-hq/pull/8156/files#diff-3331173c6439f2ca00e3adef54c7ced6R58. I can refactor whole api to use elasticsearch if it's worth to do it.

Comment (kkrampa) [inline corehq/apps/api/domain_metadata.py]: @snopoke Ok, just ignore my comment above. I just realized that I can wrap document directly from ES.

File: corehq/apps/api/domain_metadata.py [hunk 3]
```diff
@@ -12,6 +12,8 @@
 from tastypie import fields
 from tastypie.exceptions import NotFound
 import operator
+from dimagi.utils.couch.database import iter_docs
```
Comment (dimagimon) [inline corehq/apps/api/domain_metadata.py]: F401 'iter_docs' imported but unused

File: corehq/apps/api/domain_metadata.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/api/domain_metadata.py]: @kkrampa I believe that the documents you get from elasticsearch will be exactly the same as the ones you get from couchdb. (The whole doc is copied into ES.) So if you just call `Domain.wrap` on the docs from ES, you'll end up with the same result.

Comment (kkrampa) [inline corehq/apps/api/domain_metadata.py]: I've just done it in that way.

**Comment by kkrampa** [comment]
@czue I'm not completely sure that is only used internally. @fyniac could you confirm that?

File: corehq/apps/api/domain_metadata.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/api/domain_metadata.py]: which I think you can get by calling `.raw_hits` instead of `.doc_ids`

**Comment by fyniac** [comment]
Yep, we're using this internally

File: corehq/apps/api/domain_metadata.py (hunk 2 above)
Comment (kkrampa) [inline corehq/apps/api/domain_metadata.py]: It's outdated diff. https://github.com/dimagi/commcare-hq/pull/8156/files#diff-3331173c6439f2ca00e3adef54c7ced6R44

File: corehq/apps/api/domain_metadata.py [hunk 4]
```diff
@@ -30,6 +31,20 @@ def get_truth(inp, relate, cut):
...
+    def __init__(self, es_query):
+        self.es_query = es_query
+
+    def count(self):
+        return self.es_query.size(0).run().total
+
+    def __getitem__(self, item):
+        if isinstance(item, slice):
+            return map(Domain.wrap, self.es_query.start(item.start).size(item.stop - item.start).run().hits)
+        raise ValueError()
```
Comment (snopoke) [inline corehq/apps/api/domain_metadata.py]: An error message here would be useful.

**Comment by kkrampa** [comment]
@dannyroberts @snopoke Updated

**Comment by snopoke** [comment]
:white_check_mark: from me once tests pass

**Comment by esoergel** [comment]
:+1: The ESQuery code looks great to me.  I like the use of an adapter class to fit it here too.

**Pull request merged**